warehouse/index-stats/src/main/java/datawave/mapreduce/shardStats/StatsHyperLogMapper.java

```diff
@@ -0,0 +1,274 @@
+            }
+            DatawaveKey dwKey = new DatawaveKey(key);
+            if (log.isTraceEnabled()) {
+                log.trace("mapper input(" + dwKey.toString() + ")");
+            }
+            
+            String fieldName = dwKey.getFieldName();
+            if (null == this.currentFieldName) {
+                this.currentFieldName = fieldName;
+            } else {
+                if (!this.currentFieldName.equals(fieldName)) {
+                    flushFieldValues(context);
+                    this.currentFieldName = fieldName;
+                }
+            }
+            
+            // add value to proper data type
+            String dataType = dwKey.getDataType();
+            DataTypeInfo typeInfo = this.dataTypeMapping.get(dataType);
+            if (null == typeInfo) {
+                // add new datatype entry
+                typeInfo = new DataTypeInfo(dwKey, this.sumUniqueCounts, this.normalPrecision, this.sparsePrecision);
+                this.dataTypeMapping.put(dataType, typeInfo);
+            }
+            typeInfo.add(dwKey.getFieldValue());
+        }
+        
+        context.progress();
+        ;
+        
+        if (log.isTraceEnabled()) {
+            log.trace("Completed map(" + key + ", " + value + ")");
+        }
+    }
+    
+    @Override
+    protected void cleanup(Context context) throws IOException, InterruptedException {
+        if (log.isTraceEnabled()) {
+            log.trace("context(" + context + ")");
+        }
+        
+        // write last enrty
+        flushFieldValues(context);
+        
+        log.info("totals keys(" + this.outputTotal + ") +  'fi' values(" + this.total + ")");
+        
+        super.cleanup(context);
+    }
+    
+    private Key generateOutputKey(final DatawaveKey parser) {
+        String row = parser.getRow().toString();
+        int idx = row.indexOf('_');
+        String date = row.substring(0, idx);
+        Text outRow = new Text(parser.getFieldName());
+        Text outFam = new Text(date);
+        Text outQual = new Text(parser.getDataType());
+        // timestamp is set by reducer
+        return new Key(outRow, outFam, outQual, this.visibility, 0);
+    }
+    
+    private void flushFieldValues(Context context) throws IOException, InterruptedException {
+        if (null != this.currentFieldName) {
+            for (DataTypeInfo entry : this.dataTypeMapping.values()) {
+                Key key = generateOutputKey(entry.key);
+                this.outputTotal++;
+                BulkIngestKey bulkKey = new BulkIngestKey(this.outputTable, key);
+                StatsHyperLogSummary sum = entry.getStatsSummary();
+                if (log.isDebugEnabled()) {
+                    log.debug("output: key(" + bulkKey + ")");
+                    log.debug("stats(" + sum.statsString() + ")");
+                }
+                
+                Value val = new Value(sum.toByteArray());
+                context.write(bulkKey, val);
```

Comment:
There is a lot of object creation in here and in generateOutputKey. The `context.write()` will serialize out the key/value pairs so the objects can actually be reused. Some candidates for reuse would be: outRow, outFam, outQual, val. Depending on how large these jobs are, the object creation could add significant processing time.

Reply:
If the object is serialized immediately then there could be some reuse. I haven't stepped into this code but I assume what you have noted is true as the BulkContextWriter.flush() method calls the same write method.

Will investigate.

Reply:
The context.write() does serialize the object. Just for fun, I wrote a simple test case to see how much additional overhead was required for the new objects versus reuse. I had to make sure that the JVM reached a steady state before these two versions would provide accurate information. I created 400000 entries for the loop and found that the overhead was roughly 2.5% (reuse ~390ms, new object ~400ms), which is not much. In the overall picture it is probably less than .01%. It is a simple change to make, so it will be done.

Reply:
Thanks for looking into it!